### DISCUSSION TITLE
Problem with the ubuntu client, windows and mac are ok

### DISCUSSION THREAD

**Discussion by rjl126** [strongswan/strongswan#433] [Q&A]
I have a strongswan vpn server which works fine for windows and mac clients, but not ubuntu.

The ubuntu client can connect successfully to the vpn, but something is wrong with ipv4 traffic, and I am not sure how to diagnose this. 
The client seems to be able to reach the internet but only has an ipv6 address (if I check whatismyipaddress.com, ipv4 is not detected) so many sites cannot be reached, and it can't reach the actual hosts to which we are restricting access with the vpn.

The VPN server is running strongSwan 5.7.2 on debian 10.
The client is running strongSwan 5.6.2 on ubuntu 18.04

Another factor is that this ubuntu computer is also sometimes used to connect to a pptp vpn and works fine on that. I notice in the output of `ip route show table all` there is a reference to this pptp vpn. I don't know much about what the route table means, but could that be interfering at all?

Thank you for any help or ideas about what is going on.

[ubuntu_client_ipsec_conf.txt](https://github.com/strongswan/strongswan/files/6603669/ubuntu_client_ipsec_conf.txt)
[ubuntu_client_syslog.txt](https://github.com/strongswan/strongswan/files/6603691/ubuntu_client_syslog.txt)
[vpn_server_ipsec_conf.txt](https://github.com/strongswan/strongswan/files/6603671/vpn_server_ipsec_conf.txt)
[vpn_server_syslog.txt](https://github.com/strongswan/strongswan/files/6603672/vpn_server_syslog.txt)

[ubuntu_client_ip_address.txt](https://github.com/strongswan/strongswan/files/6603685/ubuntu_client_ip_address.txt)
[ubuntu_client_ip_route_show_table_all.txt](https://github.com/strongswan/strongswan/files/6603686/ubuntu_client_ip_route_show_table_all.txt)
[ubuntu_client_ipsec_statusall.txt](https://github.com/strongswan/strongswan/files/6603687/ubuntu_client_ipsec_statusall.txt)
[ubuntu_client_iptables.txt](https://github.com/strongswan/strongswan/files/6603688/ubuntu_client_iptables.txt)

**Comment by rjl126** [comment]
An update: I might have solved the problem by connecting to the vpn by running `charon-cmd` instead. Still would be interested in what went wrong with my strongswan / ipsec.conf setup though.

**Reply by tobiasbrunner** [reply]
What  you describe doesn't really match the output you posted. You wrote that something is wrong with IPv4, while IPv6 works. However, there's only an IPv4 virtual address assigned to the client and the traffic selector (`10.101.0.1/32 === 0.0.0.0/0`) and route in table 220 are IPv4 only too.

You added that it works with `charon-cmd`. That shouldn't really make a difference if your config is about the same. Try comparing the log and system changes (addrs, routes etc.) to see what might be different.

**Reply by rjl126** [reply]
Sorry I didn't explain that very well. IPv6 only worked because it wasn't going through the vpn. All IPv4 traffic that was supposed to be going through the vpn wasn't working.

I notice that when I'm connected with `charon-cmd` I have this `ipsec0` interface that wasn't there when running the strongswan service 
[ubuntu_client_ip_address_charoncmd.txt](https://github.com/strongswan/strongswan/files/6612744/ubuntu_client_ip_address_charoncmd.txt)

And I also have a route to the strongswan vpn, and no mention of the pptp vpn in the route table any more 
[ubuntu_client_ip_route_show_table_all_charoncmd.txt](https://github.com/strongswan/strongswan/files/6612761/ubuntu_client_ip_route_show_table_all_charoncmd.txt)

**Reply by tobiasbrunner** [reply]
> I notice that when I'm connected with `charon-cmd` I have this `ipsec0` interface that wasn't there when running the strongswan service

OK, looks like `charon-cmd` loads the [kernel-libipsec](https://wiki.strongswan.org/projects/strongswan/wiki/kernel-libipsec) plugin that does IPsec in userland.  The regular daemon apparently doesn't load it and installs the IPsec SAs and policies in the kernel. Don't know exactly why the latter wouldn't work, you'd have to check where traffic flows and potentially gets blocked.

> And I also have a route to the strongswan vpn

That route is required to prevent traffic to the VPN gateway from getting tunneled itself via `ipsec0`. It's not necessary when IPsec is done in the kernel.

> and no mention of the pptp vpn in the route table any more

I don't see how that's related as neither client will interact with the PPTP daemon or its routes (they only modify routes in table 220). The route to the PPTP server shouldn't make a difference in the first place.

What's a bit weird with `charon-cmd` and `kernel-libipsec` are all the additional routes via `ipsec0` for your local subnets. Don't know where these come from. Maybe the log, with log level for _knl_ set to 2, could offer some explanation for it.
